File: cli/src/commands/review.ts
import ora from "ora";
import chalk from "chalk";
import { spawnDiff, chunkDiff } from "../git.js";
import { postReview } from "../api.js";
import type { ReviewResponse } from "../types.js";

export async function reviewCommand(options: {
  agents?: string;
  backend?: string;
}): Promise<void> {
  const agents = options.agents
    ? options.agents.split(",").map((a) => a.trim())
    : ["syntax", "security", "performance", "style"];

  const spinner = ora("Capturing git diff...").start();

  try {
    const rawDiff = await spawnDiff();
    if (!rawDiff) {
      spinner.warn("No uncommitted changes found.");
      return;
    }

    spinner.text = "Chunking diff...";
    const chunks = chunkDiff(rawDiff);
    spinner.info(`Found ${chunks.length} chunk(s) across files`);

    spinner.text = `Running AI review (${agents.join(", ")})...`;
    const result: ReviewResponse = await postReview(
      { diff_chunks: chunks, agents },
      options.backend
    );

    spinner.succeed("Review complete!\n");
    renderResults(result);
  } catch (err: unknown) {
    const message = err instanceof Error ? err.message : String(err);
    spinner.fail(`Review failed: ${message}`);
    process.exit(1);
  }
}

function renderResults(result: ReviewResponse): void {
  const { tech_debt_score, finding_counts, findings } = result;

  const scoreColor =
    tech_debt_score >= 80
      ? chalk.green
      : tech_debt_score >= 50
        ? chalk.yellow
        : chalk.red;

  console.log(scoreColor.bold(`\n  Tech Debt Score: ${tech_debt_score}/100\n`));

  console.log(
    `  ${chalk.red(`HIGH: ${finding_counts.high}`)}  ${chalk.yellow(`MEDIUM: ${finding_counts.medium}`)}  ${chalk.blue(`LOW: ${finding_counts.low}`)}`
  );
  console.log(`  Review ID: ${result.review_id}\n`);

  if (findings.length === 0) {
    console.log(chalk.green("  No issues found. Great work!\n"));
    return;
  }

  console.log(chalk.bold("  Findings:\n"));
  for (const f of findings) {
    const sevColor =
      f.severity === "HIGH"
        ? chalk.red
        : f.severity === "MEDIUM"
          ? chalk.yellow
          : chalk.blue;

    console.log(`  ${sevColor.bold(f.severity)} ${chalk.dim(`${f.file_path}:${f.line}`)}`);
    console.log(`    ${chalk.bold(f.title)}`);
    console.log(`    ${f.description}`);
    console.log(`    ${chalk.cyan(`Suggestion: ${f.suggestion}`)}\n`);
  }
}

File: backend/schemas/review.py
from enum import Enum

from pydantic import BaseModel


class FindingSeverity(str, Enum):
    HIGH = "HIGH"
    MEDIUM = "MEDIUM"
    LOW = "LOW"


class DiffChunk(BaseModel):
    file_path: str
    chunk_index: int
    content: str


class ReviewRequest(BaseModel):
    diff_chunks: list[DiffChunk]
    agents: list[str] = ["syntax", "security", "performance", "style"]


class Finding(BaseModel):
    agent: str
    severity: FindingSeverity
    file_path: str
    line: int
    title: str
    description: str
    suggestion: str


class FindingCounts(BaseModel):
    high: int = 0
    medium: int = 0
    low: int = 0


class ReviewResponse(BaseModel):
    review_id: str
    tech_debt_score: int
    finding_counts: FindingCounts
    findings: list[Finding]


class MetricsSummary(BaseModel):
    total_reviews: int
    current_score: int
    score_trend: str
    week_delta: int


class ReviewHistoryEntry(BaseModel):
    review_id: str
    timestamp: str
    tech_debt_score: int
    finding_counts: FindingCounts


class ReviewHistoryResponse(BaseModel):
    reviews: list[ReviewHistoryEntry]
